In [ ]:
#importing necessary libraries
import nltk
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer

In [ ]:
nltk.download('punkt') # tokenize
nltk.download('wordnet') # lemmatize
nltk.download('stopwords') #stopwords
nltk.download('averaged_perceptron_tagger') # POS

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

In [ ]:
#reading document

c1 = pd.read_csv('/content/IA2.csv', index_col=0, header= None)
c1.head()


1
0                                                   
1  This is the place to go after this branch to t...
2  The direct delivery from Star Theater in Chian...
3  It is very delicious food. Good taste Unspoilt...
4  This house is the second time. I used to be a ...
5  Featured is majestic. Located on the road 107 ...

In [ ]:
# tokenize

c1_tokenized = []
for review in c1.iloc[:,0]:
  c1_tokenized.append(nltk.word_tokenize(review))

#c1_tokenized = pd.DataFrame(c1_tokenized)
print(c1_tokenized)



[['This', 'is', 'the', 'place', 'to', 'go', 'after', 'this', 'branch', 'to', 'the', 'afternoon', ',', 'not', 'much', 'people', '.', 'Now', 'the', 'popular', 'menu', 'is', 'ice', 'crystal', '.', 'Hes', 'not', 'long', 'Feel', 'very', 'sorry', 'T', '^', 'T', 'This', 'is', 'a', 'full', 'set', 'of', 'Chocolate', 'Mud', 'Brownie', '(', '155.-', '฿', ')', 'Chocolate', 'Brownie', 'Sprinkle', 'with', 'vanilla', 'ice', 'cream', 'and', 'whipped', 'cream', '.', 'This', 'is', 'a', 'very', 'hot', 'menu', '.', '5/5', 'Go', 'to', 'Ferrero', 'Honey', 'Toast', '(', '245.-', '฿', ')', 'Honey', 'Toro', 'topped', 'with', 'nuts', 'Topping', 'up', 'is', 'Ferrero', '.', 'Fan', 'club', 'We', 'are', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', 'This', 'is', 'a', 'good', 'service', '.', 'It', '&', '#', '39', ';', 's', 'a', 'little', 'car', 'park', '.', 'If', 'the', 'parking', 'lot', 'of', 'the', 'building', 'will', 'be', 'stamped', 'the', 'first', '2', 'hrs', '20', 'b

In [ ]:
# lemmatize
lemmatizer = nltk.stem.WordNetLemmatizer()
c1_lemmatized =[]

x= len(c1_tokenized)
for review in c1_tokenized:
  t=[]
  for token in review:
    if token.isalpha():
      t.append(lemmatizer.lemmatize(token))
  c1_lemmatized.append(t)      
print(c1_lemmatized)

[['This', 'is', 'the', 'place', 'to', 'go', 'after', 'this', 'branch', 'to', 'the', 'afternoon', 'not', 'much', 'people', 'Now', 'the', 'popular', 'menu', 'is', 'ice', 'crystal', 'Hes', 'not', 'long', 'Feel', 'very', 'sorry', 'T', 'T', 'This', 'is', 'a', 'full', 'set', 'of', 'Chocolate', 'Mud', 'Brownie', 'Chocolate', 'Brownie', 'Sprinkle', 'with', 'vanilla', 'ice', 'cream', 'and', 'whipped', 'cream', 'This', 'is', 'a', 'very', 'hot', 'menu', 'Go', 'to', 'Ferrero', 'Honey', 'Toast', 'Honey', 'Toro', 'topped', 'with', 'nut', 'Topping', 'up', 'is', 'Ferrero', 'Fan', 'club', 'We', 'are', 'This', 'is', 'a', 'good', 'service', 'It', 's', 'a', 'little', 'car', 'park', 'If', 'the', 'parking', 'lot', 'of', 'the', 'building', 'will', 'be', 'stamped', 'the', 'first', 'hr', 'baht', 'sit', 'a', 'long', 'time', 'But', 'if', 'you', 'park', 'other', 'building', 'it', 'will', 'be', 'But', 'it', 's', 'just', 'this', 'parking', 'lot', 'is', 'already', 'low'], ['The', 'direct', 'delivery', 'from', 'Star'

In [ ]:
# remove stop words
from nltk.corpus import stopwords

c1_stop =[]

for review in c1_lemmatized:
  t=[token for token in review if not token.lower() in stopwords.words('english')]
  c1_stop.append(t)      
print(c1_stop)

[['place', 'go', 'branch', 'afternoon', 'much', 'people', 'popular', 'menu', 'ice', 'crystal', 'Hes', 'long', 'Feel', 'sorry', 'full', 'set', 'Chocolate', 'Mud', 'Brownie', 'Chocolate', 'Brownie', 'Sprinkle', 'vanilla', 'ice', 'cream', 'whipped', 'cream', 'hot', 'menu', 'Go', 'Ferrero', 'Honey', 'Toast', 'Honey', 'Toro', 'topped', 'nut', 'Topping', 'Ferrero', 'Fan', 'club', 'good', 'service', 'little', 'car', 'park', 'parking', 'lot', 'building', 'stamped', 'first', 'hr', 'baht', 'sit', 'long', 'time', 'park', 'building', 'parking', 'lot', 'already', 'low'], ['direct', 'delivery', 'Star', 'Theater', 'Chiang', 'Mai', 'guarantee', 'really', 'cool', 'variety', 'flavor', 'choose', 'baht', 'per', 'piece', 'taste', 'choose', 'buy', 'buy'], ['delicious', 'food', 'Good', 'taste', 'Unspoilt', 'people', 'like', 'eat', 'spicy', 'much', 'cooked', 'shop', 'ha', 'prepared', 'Chili', 'Sauce', 'Chili', 'sauce', 'tomato', 'service', 'ok'], ['house', 'second', 'time', 'used', 'pig', 'pan', 'seafood', 'P

In [ ]:
#converting the right format for further processing

master = [' '.join(review) for review in c1_stop]
print(master)

['place go branch afternoon much people popular menu ice crystal Hes long Feel sorry full set Chocolate Mud Brownie Chocolate Brownie Sprinkle vanilla ice cream whipped cream hot menu Go Ferrero Honey Toast Honey Toro topped nut Topping Ferrero Fan club good service little car park parking lot building stamped first hr baht sit long time park building parking lot already low', 'direct delivery Star Theater Chiang Mai guarantee really cool variety flavor choose baht per piece taste choose buy buy', 'delicious food Good taste Unspoilt people like eat spicy much cooked shop ha prepared Chili Sauce Chili sauce tomato service ok', 'house second time used pig pan seafood Price Baht mall use kind common sea food fresh diarrhea first time eat diarrhea cockroach stink rotten eat shrimp enough review', 'Featured majestic Located road Mae Taeng restaurant cozy atmosphere Today people much green tea quietly enough dessert good enough bring piece high price place traveler stay mont cafe Buy leave f

In [ ]:
#tfidf


tfidf_master = TfidfVectorizer(ngram_range=(1, 2),min_df=3) # remove low frequency words
tfidf_master.fit(master)
tfidf_final = tfidf_master.transform(master)


#getting vocabulary
data_items = tfidf_master.vocabulary_.items()
data_list = list(data_items)
df = pd.DataFrame(data_list, columns = ['review', 'place'] )
df.sort_values('place', inplace = True)
df.set_index('place', inplace = True)


output = pd.DataFrame(tfidf_final.toarray(), columns = df['review'])
output.to_csv(r'/content/output.csv', index=False)
print(output.shape)
output.head()

(100, 1321)


review  able  absolutely  accept  accident  ...  youthful  yuan  zhang  zhou
0        0.0         0.0     0.0       0.0  ...       0.0   0.0    0.0   0.0
1        0.0         0.0     0.0       0.0  ...       0.0   0.0    0.0   0.0
2        0.0         0.0     0.0       0.0  ...       0.0   0.0    0.0   0.0
3        0.0         0.0     0.0       0.0  ...       0.0   0.0    0.0   0.0
4        0.0         0.0     0.0       0.0  ...       0.0   0.0    0.0   0.0

[5 rows x 1321 columns]

In [ ]:
#tfidf with POS tag

c2 = [' '.join(review) for review in c1_tokenized]

POS_c2 = []

for doc in c2:
    token_doc = nltk.word_tokenize(doc)
    POS_token_doc = nltk.pos_tag(token_doc)
    temp = [] # list of concatenated token with POS tag
    for i in POS_token_doc:
        temp.append(i[0] + "_" + i[1])
    POS_c2.append(" ".join(temp))

tfidf_master = TfidfVectorizer(min_df=4) # remove low frequency words
tfidf_master.fit(POS_c2)
c2_tfidf_final = tfidf_master.transform(POS_c2)


#getting vocabulary
data_items = tfidf_master.vocabulary_.items()
data_list = list(data_items)
df = pd.DataFrame(data_list, columns = ['review', 'place'] )
df.sort_values('place', inplace = True)
df.set_index('place', inplace = True)


output_pos = pd.DataFrame(c2_tfidf_final.toarray(),  columns = df['review'])
output_pos.to_csv(r'/content/output_pos.csv', index=False)
print(output_pos.shape)
output_pos.head()

(100, 907)


review  1_cd      2_cd     30_cd  ...  your_prp  yourself_prp  youth_nn
0        0.0  0.064543  0.000000  ...       0.0      0.000000       0.0
1        0.0  0.000000  0.260853  ...       0.0      0.238942       0.0
2        0.0  0.000000  0.000000  ...       0.0      0.000000       0.0
3        0.0  0.000000  0.000000  ...       0.0      0.000000       0.0
4        0.0  0.140975  0.000000  ...       0.0      0.000000       0.0

[5 rows x 907 columns]

The dimensions for the the tfidf vector in question 4 is 100 X 1321

The dimensions for the the tfidf vector in question 5 is 100 X 907